contracts/libraries/WeightedOracleLibrary.sol

```diff
@@ -0,0 +1,56 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity >=0.5.0 <0.8.0;
+
+import '@uniswap/v3-core/contracts/interfaces/IUniswapV3Pool.sol';
+
+/// @title Weighted Oracle library
+/// @notice Provides functions to integrate with different tier oracles of the same V3 pair
+library WeightedOracleLibrary {
+
+    struct TimeWeightedObservation {
+        int24 timeWeightedAverageTick;
+        uint256 timeWeightedHarmonicMeanLiquidity;
+    }
+
+    /// @notice Fetches time-weighted observations across different tiers of the same Uniswap V3 pair
+    /// @param pools Addresses of different tiers of the same Uniswap V3 pair that we want to observe
+    /// @param period Number of seconds in the past to start calculating the time-weighted observations
+    /// @return observations An array of obervations that have been time-weighted from (block.timestamp - period) to block.timestamp
+    function consult(address[] calldata pools, uint32 period) internal view returns (TimeWeightedObservation[] memory observations) {
```

Comment:
there is no guarantee in the interface these pools are for the same token pair

you may want to take `address token0, address token1, uint24[] feeAmounts` and compute pool addresses to enforce it, but that unfortunately also requires you to pass in factory address

Reply:
i know this makes testing harder, but the natspec indicates that it returns data `across different tiers of the same Uniswap V3 pair`. you can also clarify the natspec

Reply:
You are right. Now that this function only returns observations, there is no need to have the pools be tiers of the same pair. I'll update the natspec